pilot/pkg/model/push_context.go

```diff
@@ -1903,16 +1905,75 @@ func (ps *PushContext) initEnvoyFilters(env *Environment) {
 		return in < jn
 	})
 
-	ps.envoyFiltersByNamespace = make(map[string][]*EnvoyFilterWrapper)
 	for _, envoyFilterConfig := range envoyFilterConfigs {
 		efw := convertToEnvoyFilterWrapper(&envoyFilterConfig)
-		if _, exists := ps.envoyFiltersByNamespace[envoyFilterConfig.Namespace]; !exists {
-			ps.envoyFiltersByNamespace[envoyFilterConfig.Namespace] = make([]*EnvoyFilterWrapper, 0)
-		}
 		ps.envoyFiltersByNamespace[envoyFilterConfig.Namespace] = append(ps.envoyFiltersByNamespace[envoyFilterConfig.Namespace], efw)
 	}
 }
 
+// pre computes envoy filters per namespace
+func (ps *PushContext) updateEnvoyFilters(env *Environment, changedEnvoyFilters []ConfigKey, previousIndex map[string][]*EnvoyFilterWrapper) {
```

Comment:
In this way, we will need to handle add/update and delete. This is a little too complex, I am concerned it will not improve when the number of changed envoyfiter is large

Reply:
for people with little amount of EnvoyFilters, which is probably much more common, the difference should be neglegible.
for poeple with a huge amount of EnvoyFilters, it's probably impossible to update a huge amount of them in a single debounce window.

I really don't see any realistic scenario where this changes could end up degrading performance instead of improving it, but i can feature flag this behaviour if that sounds better.

Reply:
One approach I was thinking of was to combine this with `initEnvoyFilters`. allow it to take in `prev` and `updated`.

Pseudo code:

```go
now := env.List(EnvoyFilter)
res := make([]config, 0, len(now))
for _, cfg := range now {
	if changed.Contains(cfg) {
		// It changed, recompute
		res = append(res, computeEnvoyFilter(cfg))
	} else if prev.Contains(cfg) {
		// Not changed and we already have it, grab the previous one
		res = append(res, prev.Get(cfg))
	} else {
		// New config, compute it
		res = append(res, computeEnvoyFilter(cfg))
	}
}
```

When we currently call initEnvoyFilters we just pass nil for each of the others. This would involve converting `changed` and `prev` to a map of some sort so we can do a lookup, but seems a bit simpler?

Reply:
I was thinking that probably Listing all envoy filters all the time is probably much more expensive than just specifically getting the required envoyfilters, `List` allocates a big slice to hold all envoyfilters. Plus the conversions of `changed` and `prev`.

I could definitely take that approach, just wondering if it's not better to keep it a bit more complicated, but more efficient. I could probably separate the current logic in some methods to keep it simpler.

Reply:
I have some doubts the `List` call is so bad. Is it a guess or do we observe that?

The reason I ask is we do a List() call on basically everything, so I would expect it to blow up much sooner. Allocating a list of 2k elements, which is basically what List does, is pretty cheap I would think

Reply:
It's really a guess, i know it's cheap to allocate that amount of elements, but it will get bigger eventually, and it's probably putting some unnecessary stress on the GC.
I wouldn't take 2k elements as a strict number, we will probably have much more, unless we can incorporate some of our needs in Istio APIs.

Reply:
I did a quick benchmark, 0.2ms to allocate 200k size slice. if you get anywhere near that limit things are going to blow up for other reasons anyways. can we please switch to this approach? Its just as efficient, consistent with the rest of the code, and minimizes risk and complexity.

Reply:
Its one list call for all namespaces in my recommendation - as we do today - not per namespace

Reply:
yes, but the approach you suggest still involves rebuilding the whole index, we would really only avoid computing the envoyfilter, it's definitely better than what we currently do, but I don't think it's as efficient as copying the entire previous index and only doing the necessary changes to it.